aptos-move/mvhashmap/src/versioned_code.rs

```diff
@@ -0,0 +1,214 @@
+    }
+
+    fn read(
+        &self,
+        key: &K,
+        txn_idx: TxnIndex,
+    ) -> anyhow::Result<(Arc<V>, Option<HashValue>), MVCodeError> {
+        use MVCodeError::*;
+
+        match self.values.get(key) {
+            Some(v) => {
+                if let Some((idx, entry)) = v.versioned_map.range(0..txn_idx).next_back() {
+                    let flag = entry.flag();
+                    if flag == FLAG_ESTIMATE {
+                        // Found a dependency.
+                        return Err(Dependency(*idx));
+                    }
+                    // The entry should be populated.
+                    debug_assert!(flag == FLAG_DONE);
+
+                    Ok((entry.module.clone(), entry.maybe_hash))
+                } else {
+                    Err(NotFound)
+                }
+            },
+            None => Err(NotFound),
+        }
+    }
+
+    pub(crate) fn write(&self, key: &K, txn_idx: TxnIndex, data: V) {
+        let mut v = self.values.entry(key.clone()).or_default();
+        v.versioned_map.insert(
+            txn_idx,
+            CachePadded::new(Entry::new_write_from(FLAG_DONE, data)),
+        );
+    }
+
+    pub(crate) fn store_executable(
+        &self,
+        key: &K,
+        descriptor: ExecutableDescriptor,
+        executable: X,
+    ) {
+        let x = Arc::new(executable);
+        match descriptor {
+            ExecutableDescriptor::Published(hash) => {
+                let mut v = self.values.get_mut(key).expect("Path must exist");
+                v.executables.entry(hash).or_insert(x);
+            },
+            ExecutableDescriptor::Storage => {
+                let mut v = self.values.entry(key.clone()).or_default();
+                v.base_executable.get_or_insert(x);
+            },
+        };
+    }
+
+    pub(crate) fn fetch_module(
+        &self,
+        key: &K,
+        txn_idx: TxnIndex,
+    ) -> anyhow::Result<MVCodeOutput<V, X>, MVCodeError> {
+        use MVCodeError::*;
+        use MVCodeOutput::*;
+
+        match self.read(key, txn_idx) {
+            Ok((module, maybe_hash)) => {
+                let v = self.values.get(key).expect("Path must exist");
+
+                Ok(match maybe_hash {
+                    Some(hash) => {
+                        match v.executables.get(&hash) {
+                            Some(x) => Executable((x.clone(), false)), // from the block
+                            None => Module((module, maybe_hash)),
+                        }
+                    },
+                    None => Module((module, None)), // Deletion
+                })
+            },
+            Err(NotFound) => {
+                match self.values.get(key) {
+                    Some(v) => v
+                        .base_executable
+                        .as_ref()
+                        .map(|x| Executable((x.clone(), true))) // storage-version
+                        .ok_or(NotFound),
+                    None => Err(NotFound),
```

Comment:
When do you store base_executable? Should not you read from storage now and store? Or the wrapper does it and write back the base executable?

Reply:
the VM is the only one that can produce executables, so storage or not it will go through the "wrapper" and store the executable. Storage version doesn't change so no need for the hash as key.